File: src/CannabisMath.Core/Dosing/DoseCalculator.cs
namespace CannabisMath.Core.Dosing;

public static class DoseCalculator
{
    public static decimal MgPerServing(decimal totalMg, decimal servings)
    {
        ValidateNonNegative(totalMg, nameof(totalMg));
        ValidatePositive(servings, nameof(servings));

        return totalMg / servings;
    }

    public static decimal TotalMgFromPercent(decimal weightGrams, decimal potencyPercent)
    {
        ValidateNonNegative(weightGrams, nameof(weightGrams));
        ValidateNonNegative(potencyPercent, nameof(potencyPercent));

        return weightGrams * 1000m * (potencyPercent / 100m);
    }

    public static decimal MgPerGram(decimal potencyPercent)
    {
        ValidateNonNegative(potencyPercent, nameof(potencyPercent));

        return 1000m * (potencyPercent / 100m);
    }

    public static decimal DabMg(decimal potencyPercent, decimal dabSizeGrams)
    {
        ValidateNonNegative(potencyPercent, nameof(potencyPercent));
        ValidateNonNegative(dabSizeGrams, nameof(dabSizeGrams));

        return TotalMgFromPercent(dabSizeGrams, potencyPercent);
    }

    private static void ValidateNonNegative(decimal value, string parameterName)
    {
        if (value < 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value cannot be negative.");
        }
    }

    private static void ValidatePositive(decimal value, string parameterName)
    {
        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value must be greater than zero.");
        }
    }
}

File: js/src/dosing/doseCalculator.ts
export class DoseCalculator {
  static mgPerServing(totalMg: number, servings: number): number {
    this.validateNonNegative(totalMg, "totalMg");
    this.validatePositive(servings, "servings");

    return totalMg / servings;
  }

  static totalMgFromPercent(weightGrams: number, potencyPercent: number): number {
    this.validateNonNegative(weightGrams, "weightGrams");
    this.validateNonNegative(potencyPercent, "potencyPercent");

    return weightGrams * 1000 * (potencyPercent / 100);
  }

  static mgPerGram(potencyPercent: number): number {
    this.validateNonNegative(potencyPercent, "potencyPercent");

    return 1000 * (potencyPercent / 100);
  }

  static dabMg(potencyPercent: number, dabSizeGrams: number): number {
    this.validateNonNegative(potencyPercent, "potencyPercent");
    this.validateNonNegative(dabSizeGrams, "dabSizeGrams");

    return this.totalMgFromPercent(dabSizeGrams, potencyPercent);
  }

  private static validateNonNegative(value: number, name: string): void {
    if (value < 0) {
      throw new Error(`${name} cannot be negative`);
    }
  }

  private static validatePositive(value: number, name: string): void {
    if (value <= 0) {
      throw new Error(`${name} must be greater than zero`);
    }
  }
}